pinot-core/src/main/java/org/apache/pinot/core/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -29,97 +30,142 @@
-          // NOTE: Update the record location when there is a tie on the timestamp because during the segment
-          //       commitment, when loading the committed segment, it should replace the old record locations in case
-          //       the order of records changed.
-
-          // Remove the doc from the valid doc ids of the previous location
-          if (v.isConsuming()) {
-            // Previous location is a consuming segment, whose valid doc ids are maintained locally. Only update the
-            // valid doc ids when the update is from the same segment.
-            if (recordLocation.isConsuming() && recordLocation.getSegmentName().equals(v.getSegmentName())) {
-              validDocIds.remove(v.getDocId());
-            }
-          } else {
-            ThreadSafeMutableRoaringBitmap validDocIdsForPreviousLocation =
-                _segmentToValidDocIdsMap.get(v.getSegmentName());
-            if (validDocIdsForPreviousLocation != null) {
-              validDocIdsForPreviousLocation.remove(v.getDocId());
-            } else {
-              LOGGER.warn("Failed to find valid doc ids for previous location: {}", v.getSegmentName());
-            }
-          }
-
-          validDocIds.add(recordLocation.getDocId());
-          return recordLocation;
+        // Update the record location when getting a newer timestamp
+        if (recordInfo._timestamp > currentRecordLocation.getTimestamp()) {
+          currentRecordLocation.getValidDocIds().remove(currentRecordLocation.getDocId());
+          validDocIds.add(recordInfo._docId);
+          return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
         } else {
-          // No need to update
-          return v;
+          return currentRecordLocation;
         }
       } else {
         // New primary key
-        validDocIds.add(recordLocation.getDocId());
-        return recordLocation;
+        validDocIds.add(recordInfo._docId);
+        return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
       }
     });
   }
 
   /**
-   * Removes the upsert metadata for the given segment.
+   * Removes the upsert metadata for the given immutable segment. No need to remove the upsert metadata for the
+   * consuming segment because it should be replaced by the committed segment.
    */
-  public void removeSegment(String segmentName) {
+  public synchronized void removeSegment(String segmentName, ThreadSafeMutableRoaringBitmap validDocIds) {
     LOGGER.info("Removing upsert metadata for segment: {}", segmentName);
-    _primaryKeyToRecordLocationMap.forEach((k, v) -> {
-      if (v.getSegmentName().equals(segmentName)) {
-        // NOTE: Check and remove to prevent removing the key that is just updated.
-        _primaryKeyToRecordLocationMap.remove(k, v);
-      }
-    });
-    _segmentToValidDocIdsMap.remove(segmentName);
+
+    if (!validDocIds.getMutableRoaringBitmap().isEmpty()) {
```

Comment:
does this check the case that a replaced segment shall not remove the keys of the newly loaded? Perhaps we shall consider a state of tracking the current segmentImpl (and its corresponding `validDocIds`) for a segment name?

Reply:
Yes, it removes the entry only when the reference of the valid doc ids are the same. It won't touch the upsert metadata for other segments.